streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala

```diff
@@ -212,6 +212,86 @@ class ReceiverSuite extends TestSuiteBase with Timeouts with Serializable {
+  def testBlockGeneratorCongestion(congestionStrategy: CongestionStrategy,
+                                   blockIntervalMs: Int,
+                                   maxRatePerBlock: Int,
+                                   expectedBlocks: Int,
+                                   errorBound: Double = 0.05): Unit = {
+    congestionStrategy.onBlockBoundUpdate(maxRatePerBlock)
+    val blockGeneratorListener = new FakeBlockGeneratorListenerWithStrategy(congestionStrategy)
+    val conf = new SparkConf().set("spark.streaming.blockInterval", s"${blockIntervalMs}ms")
+    val blockGenerator = new BlockGenerator(blockGeneratorListener, 1, conf)
+
+    val waitTime = expectedBlocks * blockIntervalMs
+    val expectedMessages = maxRatePerBlock * expectedBlocks
+    val expectedMessagesPerBlock = maxRatePerBlock
+    val generatedData = new ArrayBuffer[Int]
+
+    // Generate blocks
+    val startTime = System.currentTimeMillis()
+    blockGenerator.start()
+    var count = 0
+    while(System.currentTimeMillis - startTime < waitTime) {
+      blockGenerator.addData(count)
+      generatedData += count
+      count += 1
+    }
+    blockGenerator.stop()
+
+    val recordedBlocks = blockGeneratorListener.arrayBuffers
+    val recordedData = recordedBlocks.flatten
+    assert(blockGeneratorListener.arrayBuffers.size > 0, "No blocks received")
+
+    // recordedData size should be close to the expected rate; use an error margin proportional to
+    // the value, so that rate changes don't cause a brittle test
+    val minExpectedMessages = expectedMessages - errorBound * expectedMessages
+    val maxExpectedMessages = expectedMessages + errorBound * expectedMessages
+    val numMessages = recordedData.size
+    assert(
+      numMessages >= minExpectedMessages && numMessages <= maxExpectedMessages,
+      s"#records received = $numMessages, not between $minExpectedMessages and $maxExpectedMessages"
+    )
+
+    // XXX Checking every block would require an even distribution of messages across blocks,
+    // which throttling code does not control. Therefore, test against the average.
+    val minExpectedMessagesPerBlock = expectedMessagesPerBlock -
+      errorBound * expectedMessagesPerBlock
+    val maxExpectedMessagesPerBlock = expectedMessagesPerBlock
+      + errorBound * expectedMessagesPerBlock
+    val receivedBlockSizes = recordedBlocks.map { _.size }.mkString(",")
+
+    // the first and last block may be incomplete, so we slice them out
+    val validBlocks = recordedBlocks.drop(1).dropRight(1)
+    val averageBlockSize = validBlocks.map(block => block.size).sum / validBlocks.size
+
+    assert(
+      averageBlockSize >= minExpectedMessagesPerBlock &&
+        averageBlockSize <= maxExpectedMessagesPerBlock,
+      s"# records in received blocks = [$receivedBlockSizes], not between " +
+        s"$minExpectedMessagesPerBlock and $maxExpectedMessagesPerBlock, on average"
+    )
+  }
+
+  ignore("block generator drop"){
```

Comment:
Long running time. Here's the revertible commit that does it: https://github.com/huitseeker/spark/commit/cf6b139433d13d6607bcb1069d56ec8e8efeff69